backend/test_observer/external_apis/c3/c3.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import os
+from collections.abc import Iterable
+
+import requests
+from requests import Request, Response
+
+from .models import Report, SubmissionStatus
+
+logger = logging.getLogger("test-observer-backend")
+
+
+class C3Api:
+    def __init__(self):
+        self._client_id = os.environ.get("C3_CLIENT_ID", "")
+        self._client_secret = os.environ.get("C3_CLIENT_SECRET", "")
+        self._bearer_token = ""
+
+    def get_submissions_statuses(
+        self, ids: Iterable[int]
+    ) -> dict[int, SubmissionStatus]:
+        str_ids = [str(status_id) for status_id in ids]
+        response = self._authenticate_and_send(
+            Request(
+                method="GET",
+                url="https://certification.canonical.com/api/v2/submissions/status/",
+                params={"id__in": ",".join(str_ids), "limit": len(str_ids)},
+            )
+        )
+
+        if response.ok:
+            statuses = response.json()["results"]
+            return {json["id"]: SubmissionStatus(**json) for json in statuses}
+        else:
+            logger.warning(response.text)
+            return {}
+
+    def get_reports(self, ids: Iterable[int]) -> dict[int, Report]:
+        str_ids = [str(report_id) for report_id in ids]
+        response = self._authenticate_and_send(
+            Request(
+                method="GET",
+                url="https://certification.canonical.com/api/v2/reports/summary/",
+                params={"id__in": ",".join(str_ids), "limit": len(str_ids)},
+            )
+        )
+
+        if response.ok:
+            reports = response.json()["results"]
+            return {json["id"]: Report(**json) for json in reports}
+        else:
+            logger.warning(response.text)
+            return {}
+
+    def _authenticate_and_send(self, request: Request) -> Response:
+        prepared_request = request.prepare()
+        prepared_request.headers["Authorization"] = f"Bearer {self._bearer_token}"
+        session = requests.session()
+        response = session.send(prepared_request)
+        if response.status_code in (401, 403):
+            self._authenticate()
```

Comment:
It seems like this if-condition checks whether the authorization token is still valid and refreshes it, should we maybe extract this into a helper function `_refresh_and_resend_request()` or a similar name to make it a little bit more clear on why we are resending the request?